[DOCUMENTATION] Prototype documentation build glue (#110)
reactive-firewall/multicast: docs/FAQ.md

(assuming project is setup and installed and you want to listen on 0.0.0.0)

```bash
# cd /MY-AWSOME-DEV-PATH
python3 -m multicast HEAR --use-std --port 59595 --join-mcast-groups 224.0.0.1 --bind-group 224.0.0.1
```

Caveat: RCEV is much more usefull if actually used in a loop like:

```bash
# cd /MY-AWSOME-DEV-PATH
while true ; do # unitl user ctl+c inturupts
python3 -m multicast RECV --use-std --port 59595 --join-mcast-groups 224.0.0.1 --bind-group 224.0.0.1
done
```

### How do I use this to send UDP Multicast?

(assuming project is setup and installed)

```bash
# cd /MY-AWSOME-DEV-PATH
python3 -m multicast SAY --mcast-group 224.1.1.2 --port 59595 --message "Hello World!"
```

### What is the basic API via python (instead of bash like above):

#### Caveat: this module is still a BETA
[Here is how it is tested right now](https://github.com/reactive-firewall/multicast/blob/cdd577549c0bf7c2bcf85d1b857c86135778a9ed/tests/test_usage.py#L251-L554)

```python3
import multicast as multicast
from multiprocessing import Process as Process

# setup some stuff
_fixture_PORT_arg = int(59595)
_fixture_mcast_GRP_arg = """224.0.0.1"""  # only use dotted notation for multicast group addresses

> _fixture_host_BIND_arg

Author: 🙈 Oops. This PR/MR is showing my human error 🙉

```suggestion
_fixture_host_BIND_arg = """224.0.0.1"""
```

Can you please:
* read the latest "master" branch copy of "multicast/__main__.py" focusing on the argument parsing of "--join-mcast-groups" and "--bind-group"
* read the latest "master" branch copy of "multicast/__init__.py" focusing on the argument parsing of the class "mtool"
* read the latest "master" branch copy of "multicast/HEAR.py" focusing on the argument parsing of "--join-mcast-groups" and "--bind-group" and "HEAR"
* read [feature-documentation-sphinx](https://github.com/reactive-firewall/multicast/tree/feature-documentation-sphinx) branch copy of "docs/FAQ.md" considering it is a draft about the same `multicast` module code.

* finally, is the FAQ's example around line 70 even correct, you may assume the FAQ.md document is neither official, nor a source-of-truth, nor a record-of-truth for this. You may also consider corrected syntax-errors by reasonable guessing as needed to make a determination in this case. I look forward to your response.

_fixture_HEAR_args = [
	"""--port""", _fixture_PORT_arg,
	"""--join-mcast-groups""", _fixture_mcast_GRP_arg,

> 	"""--bind-group""", _fixture_mcast_GRP_arg"

Author: ```suggestion
	"""--bind-group""", _fixture_host_BIND_arg
```

]

# spwan a listening proc

> def inputHandle()

Author: ```suggestion
def inputHandler():
```

> 	buffer_string = str("""""")
> 	buffer_string += multicast.recv.hearstep([_fixture_mcast_GRP_arg], _fixture_PORT_arg, _fixture_host_BIND_arg, _fixture_mcast_GRP_arg)
> 	return buffer_string

Author: assume Tabs instead of spaces (as is consistent for the rest of the project)
```suggestion
	test_RCEV = multicast.recv.McastRECV()
	buffer_string = str("""""")
	buffer_string += test_RCEV._hearstep([_fixture_mcast_GRP_arg], _fixture_PORT_arg, _fixture_host_BIND_arg, _fixture_mcast_GRP_arg)
	return buffer_string
```

def printLoopStub(func):
	for i in range( 0, 5 ):
		print( str( func() ) )

p = Process(
				target=multicast.__main__.McastDispatch().doStep,
				name="HEAR", args=("HEAR", _fixture_HEAR_args,)

> 			)

Author: Consider it a convention to indent bash/python/shell/python3/make code-blocks with Hard tabs in markdown and github-markdown for the whole "multicast" project when using the Markdownlint tool. I prefer to use 1 tab in place of 4 spaces for Python and Bourne shells source-code. Let's leave the [holy war](https://softwareengineering.stackexchange.com/questions/57/tabs-versus-spaces-what-is-the-proper-indentation-character-for-everything-in-e) be, and keep it consistent with the project.

p.start()

# ... probably will return with nothing outside a handler function in a loop
```
and elsewhere (like another function or even module) for the sender:
```python3

> 

Author: better how? we've technically dropped python2.7 and it is EOL. Explain in the context of GH markdown what is actually different between the identifiers. I'm skeptical, so use a formal academic tone to explain.

> Arbitrary port to use by default, though any dynamic and free port would work.

> :exclamation: Caution: it is best to specify the port in use at this time as the default has yet to be properly assigned ( see related reactive-firewall/multicast#62 )

### What does exit code _x_ mean?

#### Python function return code meanings

`0` is the default and implies *success*, and means the process has essentially (or actually) returned nothing (or `None`)

`1` is used when a *single* result is returned (caveat: functions may return a single `tuple` instead of `None` to indicate exit code `1` by returning a `boolean` success value, and result (which may also be encapsulated as an iteratable if needed) )

`2` is used to indicate a *value and reason* are returned (caveat: functions may return a single `tuple` with a single value and reason and the value can be a `tuple`)

`-1` is used to mean *many* of unspecified length and otherwise functions as `1`

#### CLI exit code meanings

`0` *success*

`1` *none-sucsess* - and is often accompanied by warnings or errors

`2 >` *failure* of specific reason

#### Everything Else

> _(extra exit code meanings)_

Author: this is intended to be a caption instead of a heading. Have any other suggestions for this case? why or why not?

_fixture_HEAR_args = [
	"""--port""", _fixture_PORT_arg,
	"""--join-mcast-groups""", _fixture_mcast_GRP_arg,
	"""--bind-group""", _fixture_host_BIND_arg
]

# spwan a listening proc

def inputHandler():
	test_RCEV = multicast.recv.McastRECV()
	buffer_string = str("""""")
	buffer_string += test_RCEV._hearstep([_fixture_mcast_GRP_arg], _fixture_PORT_arg, _fixture_host_BIND_arg, _fixture_mcast_GRP_arg)
	return buffer_string

def printLoopStub(func):
	for i in range( 0, 5 ):
		print( str( func() ) )

p = Process(
				target=multicast.__main__.McastDispatch().doStep,
				name="HEAR", args=("HEAR", _fixture_HEAR_args,)
			)
p.start()

# ... probably will return with nothing outside a handler function in a loop
```
and elsewhere (like another function or even module) for the sender:
```python3

# assuming already did 'import multicast as multicast'

_fixture_SAY_args = [
	"""--port""", _fixture_PORT_arg,
	"""--mcast-group""", _fixture_mcast_GRP_arg,
	"""--message""", """'test message'"""
]
try:
	multicast.__main__.McastDispatch().doStep("SAY", _fixture_SAY_args)
	# Hint: use a loop to repeat or different arguments to varry message.
except Exception:
	p.join()

> 	raise RuntimeException("multicast seems to have failed, blah, blah")

Author: ```suggestion
	raise RuntimeError("multicast seems to have failed.")
```

### What are the defaults?

#### The default multicast group address is 224.0.0.1

From the [documentation](https://github.com/reactive-firewall/multicast/blob/v1.4/multicast/__init__.py#L185-L187):
> The Value of "224.0.0.1" is chosen as a default multicast group as per RFC-5771
> on the rational that this group address will be treated as a local-net multicast
> (caveat: one should use link-local for ipv6)

#### The default multicast Time-to-Live (TTL) is 1

From [RFC-1112 §6.1](https://www.rfc-editor.org/rfc/rfc1112#section-6.1)
> ... If the
> upper-layer protocol chooses not to specify a time-to-live, it should
> default to 1 for all multicast IP datagrams, so that an explicit
> choice is required to multicast beyond a single network.

From the [documentation](https://github.com/reactive-firewall/multicast/blob/v1.4/multicast/__init__.py#L214-L217):
> A Value of 1 (one TTL) is chosen as per [RFC-1112 §6.1](https://www.rfc-editor.org/rfc/rfc1112#section-6.1) on the rational that an
> explicit value that could traverse beyond the local connected network should be

> > chosen by the caller rather than the default value. This is inline with the principle

Author: ```suggestion
> chosen by the caller rather than the default value. This is in line with the principle
```

_fixture_HEAR_args = [
	"""--port""", _fixture_PORT_arg,
	"""--join-mcast-groups""", _fixture_mcast_GRP_arg,
	"""--bind-group""", _fixture_host_BIND_arg
]

# spawn a listening proc

def inputHandler():
	test_RCEV = multicast.recv.McastRECV()
	buffer_string = str("""""")
	buffer_string += test_RCEV._hearstep([_fixture_mcast_GRP_arg], _fixture_PORT_arg, _fixture_host_BIND_arg, _fixture_mcast_GRP_arg)
	return buffer_string

def printLoopStub(func):
	for i in range( 0, 5 ):
		print( str( func() ) )

p = Process(
				target=multicast.__main__.McastDispatch().doStep,
				name="HEAR", args=("HEAR", _fixture_HEAR_args,)
			)
p.start()

# ... probably will return with nothing outside a handler function in a loop
```
and elsewhere (like another function or even module) for the sender:
```python3

# assuming already did 'import multicast as multicast'

_fixture_SAY_args = [
	"""--port""", _fixture_PORT_arg,
	"""--mcast-group""", _fixture_mcast_GRP_arg,
	"""--message""", """'test message'"""
]
try:
	multicast.__main__.McastDispatch().doStep("SAY", _fixture_SAY_args)
	# Hint: use a loop to repeat or different arguments to vary message.
except Exception:
	p.join()

> 	raise RuntimeError("multicast seems to have failed.")

Author: this example assumes this condition is elsewhere in the code (ie from the previous example)
